### DISCUSSION TITLE
posts/Go/package

### DISCUSSION THREAD

**Discussion by Q1mi** [Q1mi/BlogComments#362] [Announcements]
https://www.liwenzhou.com/posts/Go/11_package/ 

 李文周的Blog Go语言 包 package 包 代码组织结构

**Comment by jicki** [comment]
这个 init 顺序. 为什么我执行以后不是 C 先执行 init 函数, 而是 A 先执行 init 函数呢

**Comment by 272909106** [comment]
package main

import "fmt"
import c "clac"

func main() {
	var a, b, sum, min, mul, divide int
	a = 10
	b = 11
	e := 12
	f := 13
	sum = c.Jia(a, b)
	min = c.Jian(a, b)
	mul = c.Cheng(a, b)
	divide = c.Chu(e, f)
	fmt.Println(min)
	fmt.Println(sum)
	fmt.Println(mul)
	fmt.Println(divide)
}

**Comment by asd6335683** [comment]
> @272909106
> package main
> 
> import "fmt"
> import c "clac"
> 
> func main() {
> 	var a, b, sum, min, mul, divide int
> 	a = 10
> 	b = 11
> 	e := 12
> 	f := 13
> 	sum = c.Jia(a, b)
> 	min = c.Jian(a, b)
> 	mul = c.Cheng(a, b)
> 	divide = c.Chu(e, f)
> 	fmt.Println(min)
> 	fmt.Println(sum)
> 	fmt.Println(mul)
> 	fmt.Println(divide)
> }

没有考虑浮点数

**Comment by Q1mi** [comment]
> @jicki
> 这个 init 顺序. 为什么我执行以后不是 C 先执行 init 函数, 而是 A 先执行 init 函数呢

注意import导入的顺序

**Comment by Bit0r** [comment]
问题的`clac`是不是错了，应该是`calc`？

**Comment by DrakenZeng** [comment]
咨询一下  我用 go module管理 怎么做到导入自己写的包  不知道为啥 一直报错
build command-line-arguments: cannot find module for path _/home/dark/code/goProject/iris/hero/result/Result

**Comment by Janver-wei** [comment]
一个包多次导入，会怎么样？比如 ，a包导入c包,b包也导入c包。

**Comment by Q1mi** [comment]
> @Bit0r
> 问题的`clac`是不是错了，应该是`calc`？

嗯， 是我单词写错了。。。应该是calc。感谢反馈

**Comment by Q1mi** [comment]
> @Darkells
> 咨询一下  我用 go module管理 怎么做到导入自己写的包  不知道为啥 一直报错
> build command-line-arguments: cannot find module for path _/home/dark/code/goProject/iris/hero/result/Result

我最近的一篇博客里有总结导入本地包：https://www.liwenzhou.com/posts/Go/import_local_package_in_go_module/，你可以看一下

**Comment by Q1mi** [comment]
> @Alax-W
> 一个包多次导入，会怎么样？比如 ，a包导入c包,b包也导入c包。

这个问题很常见，比如我们会在一个项目中的多个文件中都import "fmt"。

**Comment by 865077695** [comment]
一个文件夹下几个文件都是同一个package，首行用同样的package名吗？下一级文件夹又是另外的package吗？
比如a/b/  a文件夹中的几个文件都必须是相同的package,b文件夹下也必须是相同的package且不是a

**Comment by brianchiang-tw** [comment]
calc.go in /calc

```
package calc

import (
	"fmt"
)

func Add(x, y int) int {
	z := x + y
	return z
}

func Subtract(x, y int) int {
	z := x - y
	return z
}

func Multiply(x, y int) int {
	z := x * y
	return z
}

func Divide(x, y int) int {

	defer func() {
		if err := recover(); err != nil {
			fmt.Println("\nSomething wrong!!")
			fmt.Println("Error", err)
		}
		
	}()

	if y == 0 {
		panic("Divion by zero is not allowed")
	} else {
		z := x / y
		return z
	}
}
```

---

go.mod in /calc
```
module test

go 1.16

```

---

main.go in /snow

```
package main

import (
	c "calc"
	"fmt"
)

func main() {

	var result int

	result = c.Add(6, 3)
	fmt.Println("6 + 3 =", result)

	result = c.Subtract(6, 3)
	fmt.Println("6 - 3 =", result)

	result = c.Multiply(6, 3)
	fmt.Println("6 * 3 =", result)

	result = c.Divide(6, 3)
	fmt.Println("6 / 3 =", result)

	result = c.Divide(6, 0)
	fmt.Println("6 / 0 =", result)

}

```

---

go.mod in /snow

```
module mypackage

go 1.16

require "calc" v0.0.0
replace "calc" => "../calc"
```

---

output result
```
6 + 3 = 9
6 - 3 = 3
6 * 3 = 18
6 / 3 = 2

Something wrong!!
Error Divion by zero is not allowed
6 / 0 = 0
```

**Comment by yamihao** [comment]
```golang
package calc

var Calc = New()

type calc struct{}

func New() *calc {
	return &calc{}
}

func (c *calc) Add(a, b float64) float64 {
	return a + b
}

func (c *calc) Sub(a, b float64) float64 {
	return a - b
}

func (c *calc) Div(a, b float64) float64 {
	if b == 0 {
		panic("b param error")
	}
	return a / b
}

func (c *calc) Mul(a, b float64) float64 {
	return a * b
}

```

**Comment by lanmo97** [comment]
> @Q1mi
> > @Darkells
> > 咨询一下  我用 go module管理 怎么做到导入自己写的包  不知道为啥 一直报错
> > build command-line-arguments: cannot find module for path _/home/dark/code/goProject/iris/hero/result/Result
> 
> 我最近的一篇博客里有总结导入本地包：https://www.liwenzhou.com/posts/Go/import_local_package_in_go_module/，你可以看一下

博主您好，这个链接404了

**Comment by Q1mi** [comment]
> @iBroProJeff
> > @Q1mi
> > > @Darkells
> > > 咨询一下  我用 go module管理 怎么做到导入自己写的包  不知道为啥 一直报错
> > > build command-line-arguments: cannot find module for path _/home/dark/code/goProject/iris/hero/result/Result
> > 
> > 我最近的一篇博客里有总结导入本地包：https://www.liwenzhou.com/posts/Go/import_local_package_in_go_module/，你可以看一下
> 
> 博主您好，这个链接404了

[https://www.liwenzhou.com/posts/Go/import_local_package_in_go_module/](https://www.liwenzhou.com/posts/Go/import_local_package_in_go_module/)

**Comment by QuYingying0628** [comment]
1. calc包
```go
package calc

import "fmt"

func init() {
	fmt.Println("您已成功导入了calc包！")
}

// 两个int32类型数的加法
func Add(a int32, b int32) int32 {
	return a + b
}

// 两个int32类型数的减法(前一个数减后一个数)
func Subtraction(a int32, b int32) int32 {
	return a - b
}

// 两个int32类型数的乘法,返回一个int64的数
func Multiplication(a int32, b int32) int64 {
	return int64(a) * int64(b)
}

// 两个int32类型数的除法(前一个数除后一个数)
func Division(a int32, b int32) int32 {
	//defer func() {
	//	recoverResult := recover()
	//	if recoverResult != nil {
	//		fmt.Println("恢复成功！")
	//	}
	//}()

	if b == 0 {
		panic("除数为0了")
	}
	return a / b
}
```
2. snow包
```go
package snow

import (
	"awesomeProject/pkg/calc"
	"fmt"
)

func init() {
	fmt.Println("您已成功导入了snow包！")
}

// 测试calc包中的加减乘除方法
func Test(a int32,b int32){
	// 加法
	fmt.Println(a,"和",b,"的加法为",calc.Add(a,b))

	// 减法
	fmt.Println(a,"和",b,"的减法为",calc.Subtraction(a,b))

	// 乘法
	fmt.Println(a,"和",b,"的乘法为",calc.Multiplication(a,b))

	// 除法
	fmt.Println(a,"和",b,"的除法为",calc.Division(a,b))
}
```
3. main调用
```go
        // 测试snow的方法（包的调用）
	snow.Test(1,2)
	snow.Test(1,0)
```
4. 结果
```go
您已成功导入了calc包！
您已成功导入了snow包！
1 和 2 的加法为 3
1 和 2 的减法为 -1
1 和 2 的乘法为 2
1 和 2 的除法为 0
1 和 0 的加法为 1
1 和 0 的减法为 1
1 和 0 的乘法为 0
panic: 除数为0了
```
